### PR TITLE
Streaming read support for the loose ODB backend

### PR DESCRIPTION
**Pull request by ethomson**
Change the streaming reader API a bit, to provide the type and length of the object when initializing the stream. This would be a breaking API change - except, of course, that nobody is actually using the streaming reader API since none of the backends support it. So I feel pretty good breaking this API that nobody could be using.

Add streaming reader support to the loose object ODB backend. I cleaned up a bit while I was in there to add some more tests for things like read_header which had no explicit tests (and in fact failed on some of the loose ODB test corpus), teach read_header how to cope with "packlike loose objects" (which were a weird loose object format that was attempted for a bit and will never actually be seen in the wild) and drop the odb_loose internal zstream abstraction layer that ultimately became git_zstream.

Note that this depends on #4443.

### CODE REVIEW

File: src/odb_loose.c [hunk 1]
```diff
@@ -896,11 +907,139 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
+
+	*stream_out = (git_odb_stream *)stream;
+	*len_out = hdr.size;
+	*type_out = hdr.type;
+
+done:
+	if (error < 0) {
+		git_futils_mmap_free(&stream->map);
+		git_zstream_free(&stream->zstream);
+		git_hash_ctx_cleanup(hash_ctx);
```
Comment (pks-t) [inline src/odb_loose.c]: `git_hash_ctx_cleanup` will never free the context itself. (Btw, this is why I intend on distinguishing `free` and `cleanup`-style functions :P)

Comment (pks-t) [reply src/odb_loose.c]: Also: wouldn't it be easier to just have the hash context embedded inside of the stream itself? There's no need for the pointer, I think.

File: src/odb_loose.c [hunk 2]
```diff
@@ -896,11 +907,139 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
+
+	return total;
+}
+
+static void loose_backend__readstream_free(git_odb_stream *_stream)
+{
+	loose_readstream *stream = (loose_readstream *)_stream;
+
+	git_futils_mmap_free(&stream->map);
+	git_zstream_free(&stream->zstream);
```
Comment (pks-t) [inline src/odb_loose.c]: Missing the `git_hash_ctx_cleanup` here. Also missing the `git__free` if you're not changing the hash context to instead be embedded directly.

File: src/odb_loose.c [hunk 3]
```diff
@@ -910,28 +911,29 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
 	/*
 	 * if we read more than just the header in the initial read, play
 	 * that back for the caller.
 	 */
-	if (stream->start_len && len) {
-		size_t chunk = MIN(stream->start_len, len);
-		memcpy(buffer, stream->start, chunk);
+	if (start_remain && buffer_len) {
+		size_t chunk = min(start_remain, buffer_len);
+		memcpy(buffer, stream->start + stream->start_read, chunk);
```
Comment (pks-t) [inline src/odb_loose.c]: Ah, here it's getting fixed. Did you squash that into the wrong commit by accident?

File: src/odb_loose.c [hunk 4]
```diff
@@ -910,28 +911,29 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
-		stream->start_len -= chunk;
-		len -= chunk;
+		total += chunk;
+		buffer_len -= chunk;
 	}
 
-	if (len) {
-		size_t chunk = MIN(len, INT_MAX);
+	if (buffer_len) {
+		size_t chunk = min(buffer_len, INT_MAX);
```
Comment (pks-t) [inline src/odb_loose.c]: I'm surprised we never had a name clash with this `min` macro and any variable.

File: tests/odb/loose.c [hunk 5]
```diff
@@ -55,6 +55,44 @@ static void test_read_object(object_data *data)
...
+	cl_git_pass(git_odb_open(&odb, "test-objects"));
+	cl_git_pass(git_oid_fromstr(&id, data->id));
+	cl_git_pass(git_odb_open_rstream(&stream, &tmp.len, &tmp.type, odb, &id));
+
+	remain = tmp.len;
+
+	while (remain) {
+		cl_assert((ret = git_odb_stream_read(stream, ptr, blocksize)) >= 0);
+		if (ret == 0)
+			break;
```
Comment (pks-t) [inline tests/odb/loose.c]: Hum. Shouldn't we return EOF in case the stream is at its end?

File: src/zstream.c [hunk 6]
```diff
@@ -87,9 +87,51 @@ size_t git_zstream_suggest_output_len(git_zstream *zstream)
 		return ZSTREAM_BUFFER_MIN_EXTRA;
```
Comment (pks-t) [inline src/zstream.c]: There's a typo in the commit message: `git_output` should be `get_output`.

File: src/zstream.c [hunk 7]
```diff
@@ -87,9 +87,51 @@ size_t git_zstream_suggest_output_len(git_zstream *zstream)
...
+int git_zstream_get_output_chunk(
+	void *out, size_t *out_len, git_zstream *zstream)
+{
+	size_t in_queued, in_used, out_queued;
+
+	/* set up input data */
+	zstream->z.next_in = (Bytef *)zstream->in;
+	zstream->z.avail_in = (uInt)zstream->in_len;
+
+	if ((size_t)zstream->z.avail_in != zstream->in_len) {
```
Comment (pks-t) [inline src/zstream.c]: Is this handling the case where `z.avail_in = (uInt) zstream->in_len` overflows? If so, a small comment might be worthwhile.

File: src/odb_loose.c [hunk 8]
```diff
@@ -192,67 +195,6 @@ static int parse_header(
...
-	inflateEnd(s);
-
-	if ((status != Z_STREAM_END) || (s->avail_in != 0)) {
-		giterr_set(GITERR_ZLIB, "failed to finish zlib inflation; stream aborted prematurely");
-		return -1;
-	}
-
-	return 0;
-}
-
```
Comment (pks-t) [inline src/odb_loose.c]: Yay! :)

File: src/odb_loose.c [hunk 9]
```diff
@@ -131,7 +131,10 @@ static int parse_header_packlike(
 	}
 
 	out->size = size;
-	*out_len = used;
+
+	if (out_len)
+		*out_len = used;
+
```
Comment (pks-t) [inline src/odb_loose.c]: This change doesn't look related?

File: src/odb_loose.c [hunk 10]
```diff
@@ -377,33 +413,23 @@ static int read_header_loose(git_rawobj *out, git_buf *loc)
...
-		(error = git_zstream_get_output_chunk(inflated, &inflated_len, &zs)) < 0 ||
-		(error = parse_header(&hdr, &header_len, inflated, inflated_len)) < 0)
-		goto done;
+	if (!is_zlib_compressed_data(obj))
+		error = read_header_loose_packlike(out, obj, (size_t)obj_len);
+	else
+		error = read_header_loose_standard(out, obj, (size_t)obj_len);
 
-	if (!git_object_typeisloose(hdr.type)) {
+	if (!error && !git_object_typeisloose(out->type)) {
```
Comment (pks-t) [inline src/odb_loose.c]: A goto in case of error would be a bit nicer here, I guess. Special-casing error handling just because it is the last statement only asks for small error later on when more code is being inserted after it.

File: src/odb_loose.c [hunk 11]
```diff
@@ -352,7 +355,7 @@ static int read_loose(git_rawobj *out, git_buf *loc)
 	if ((error = git_futils_readbuffer(&obj, loc->ptr)) < 0)
 		goto done;
 
-	if (!is_zlib_compressed_data((unsigned char *)obj.ptr))
+	if (!is_zlib_compressed_data((unsigned char *)obj.ptr, obj.size))
 		error = read_loose_packlike(out, &obj);
 	else
 		error = read_loose_standard(out, &obj);
```
Comment (pks-t) [inline src/odb_loose.c]: Huh. Why should we treat an object as packlike though when `obj.size < 2`? Shouldn't we error out in that case instead? Something like:

```
if (is_zlib_compressed_data(&is_compressed, obj.ptr. obj.size) < 0)
    return -1;

if (!is_compressed)
    packlike();
else
    standard();
```

File: src/odb_loose.c (hunk 1 above)
Comment (ethomson) [reply src/odb_loose.c]: I had the same thought, but I'm not sure that we can:  the stream is a public struct _but_ our hash context struct should be opaque, since it could actually be implemented by a variety of mechanisms.  I'd like to reevaluate this, though, since it would be nice to not malloc.  But I think that should be a separate PR.

File: src/odb_loose.c [hunk 12]
```diff
@@ -896,11 +907,139 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
+	if (len) {
+		size_t chunk = MIN(len, INT_MAX);
+
+		if ((error = git_zstream_get_output(buffer, &chunk, &stream->zstream)) < 0)
+			return error;
+
+		total += chunk;
+	}
+
+	return total;
```
[the remark this answers is not in the document]
Comment (ethomson) [reply src/odb_loose.c]: We behave like `read(2)` by returning `0` at EOF.

File: src/odb_loose.c [hunk 13]
```diff
@@ -896,11 +907,139 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
+	size_t decompressed, head_len;
+	obj_hdr hdr;
+	int error = 0;
+
+	assert(stream_out && len_out && type_out && _backend && oid);
+
+	backend = (loose_backend *)_backend;
+	*stream_out = NULL;
+	*len_out = 0;
+	*type_out = 0;
```
[the remark this answers is not in the document]
Comment (ethomson) [reply src/odb_loose.c]: Seems reasonable; fixed.

File: src/odb_loose.c [hunk 14]
```diff
@@ -896,11 +907,139 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
+		(error = git_zstream_init(&stream->zstream, GIT_ZSTREAM_INFLATE)) < 0 ||
+		(error = git_zstream_set_input(&stream->zstream, stream->map.data, stream->map.len)) < 0)
+		goto done;
+
+	decompressed = sizeof(head);
+
+	/*
+	* inflate the initial part of the compressed buffer in order to parse the
+	* header; read the largest header possible, then store it in the `start`
+	* field of the stream object.
```
[the remark this answers is not in the document]
Comment (ethomson) [reply src/odb_loose.c]: I fixed this comment.

File: src/odb_loose.c [hunk 15]
```diff
@@ -896,11 +907,139 @@ static int loose_backend__writestream(git_odb_stream **stream_out, git_odb_backe
...
+	*len_out = 0;
+	*type_out = 0;
+
+	stream = git__calloc(1, sizeof(loose_readstream));
+	GITERR_CHECK_ALLOC(stream);
+
+	hash_ctx = git__malloc(sizeof(git_hash_ctx));
+	GITERR_CHECK_ALLOC(hash_ctx);
+
+	if (locate_object(&object_path, backend, oid) < 0) {
```
[the remark this answers is not in the document]
Comment (ethomson) [reply src/odb_loose.c]: Fixed.

File: src/zstream.c (hunk 6 above)
Comment (ethomson) [reply src/zstream.c]: Fixed.

File: src/zstream.c (hunk 7 above)
Comment (ethomson) [reply src/zstream.c]: Updated to add a comment and an improved test.

File: tests/odb/loose.c (hunk 5 above)
Comment (ethomson) [reply tests/odb/loose.c]: `0` indicates EOF (we cannot block).

File: src/odb_loose.c (hunk 2 above)
Comment (ethomson) [reply src/odb_loose.c]: [The general ODB layer handles this](https://github.com/libgit2/libgit2/blob/master/src/odb.c#L1376).  It cleans up the base object, the implementations do not.

File: src/odb_loose.c (hunk 1 above)
Comment (ethomson) [reply src/odb_loose.c]: Corrected the missing `git__free(hash_ctx)`

**Comment by ethomson** [comment]
I think that I've taken care of your issues, @pks-t

File: tests/odb/loose.c [hunk 16]
```diff
@@ -103,6 +141,22 @@ void test_odb_loose__simple_reads(void)
 	test_read_object(&some);
 }
 
+void test_odb_loose__streaming_reads(void)
+{
+	size_t blocksizes[] = { 1, 2, 4, 16, 99, 1024, 123456789, 0 };
```
Comment (pks-t) [inline tests/odb/loose.c]: Ugh, I was really confused by the `0` blocksize here. Didn't see that it was used as a terminator, so I kept on asking myself for several minutes how we can actually fully read through the stream in that case. Maybe remove that blocksize and iterate by array size instead of by sentinel?

File: tests/odb/largefiles.c [hunk 17]
```diff
@@ -73,6 +77,48 @@ void test_odb_largefiles__streamwrite(void)
 	cl_assert_equal_oid(&expected, &oid);
 }
 
+void test_odb_largefiles__streamread(void)
```
[the remark this answers is not in the document]
Comment (ethomson) [reply tests/odb/largefiles.c]: I don't think we've come to an agreement on this, no.  I'll rearrange these, but we should hash this out at the summit.  I'm definitely not the only person who prefers to add a failing test first, so the question is if we can get a workflow that accommodates all our preferences.

File: src/odb_loose.c [hunk 18]
```diff
@@ -337,10 +348,10 @@ static int inflate_disk_obj(git_rawobj *out, git_buf *obj)
...
 	decompressed = sizeof(head);
 
-	/*
-	* inflate the initial part of the compressed buffer in order to parse the
-	* header; read the largest header possible, then push back the remainder.
-	*/
+	/* inflate the initial part of the compressed buffer in order to
+	 * parse the header; read the largest header possible, then push the
+	 * remainder into the body buffer.
+	 */
```
[the remark this answers is not in the document]
Comment (ethomson) [reply src/odb_loose.c]: lololol, i trust that your :trollface: ing with the "nicely".

File: tests/odb/loose.c (hunk 16 above)
Comment (ethomson) [reply tests/odb/loose.c]: Yeah, I think you're right; I fixed this.

**Comment by ethomson** [comment]
Thanks again, updated with your comments.

File: tests/odb/largefiles.c (hunk 17 above)
[the remark this answers is not in the document]
Comment (pks-t) [reply tests/odb/largefiles.c]: I definitly agree with you. Adding failing tests first is the best way how we can demonstrate that the next commit fixes the issue. See my pull request #4500, which tackles this exact problem

File: src/odb_loose.c (hunk 18 above)
[the remark this answers is not in the document]
Comment (pks-t) [reply src/odb_loose.c]: Hehe, I am ;)

**Comment by pks-t** [comment]
Now if only GitHub had an interface to easily view the interdiff between two versions of the PR... don't really feel like going through all of it again right now :(
